WIP: Bicker serial driver

ntd opened the pull request:
This driver supports some interesting supercap DC UPSes manufactured by [Bicker](https://www.bicker.de/en/products/ups-systems/dc-ups). I'm planning to use them in the near future on small Linux-based industrial PC (mainly amd64 or arm64).

AFAICS, there are two main issues:
1. the code does not work on big-endian platforms (see the [relevant comment](https://github.com/ntd/nut/blob/bicker/drivers/bicker_ser.c#L183-L184))
2. the way I'm using `ups.delay.shutdown` stinks: most drivers defines it in `upsdrv_initinfo` but then my `upsdrv_shutdown` function crashes when trying to access it. I followed the `usbhid-ups.c` example and put it inside `upsdrv_initups`.

jimklimov commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,379 @@
+	}
+
+	/* XXX: by default, data is stored in little-endian so,
+	 * on big-endian platforms, a byte swap should be needed */
```
Note we have the `AC_C_BIGENDIAN` macro called in `configure.ac` and later `AC_SUBST(WORDS_BIGENDIAN)`, since per https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/C-Compiler.html "The default for action-if-true is to define ‘WORDS_BIGENDIAN’. The default for action-if-false is to do nothing. And finally, the default for action-if-unknown is to abort configure and tell the installer which variable he should preset to bypass this test."

There is an example use-case in NUT libshut:
* https://github.com/networkupstools/nut/blob/master/drivers/libshut.c#L337-L350

jimklimov replied:
Thinking of it, if you know the endianness of the wire protocol words (which of the two bytes is significant), you can construct int's in driver code by bit-shifting one of the bytes by 8 bits (or multiplying one by 256) and get the 16-bit word in CPU-specific format. This pattern is actually often seen in NUT code.

ntd replied:
Should be fixed in 13ff8f52b8e22da673b43d5289c6fe41d3efdacd.

jimklimov commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,379 @@
+	if (delay > 255) {
+		upslogx(LOG_WARNING, "Shutdown delay too big: %d > 255",
+			delay);
+		delay = 255;
```
Just in case, isn't `255` some sort of magical value for the device? I've seen issues with others treating "all bits set" specially (e.g. as not available).

ntd replied:
The manual just says "min 0 max 255". I used 255 and the UPS behaves as usual, i.e. it ignores the delay and cuts the output after 2 seconds.

jimklimov commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,379 @@
+	ser_flush_io(upsfd);
+
+	if (data != NULL) {
+		if (datalen > 252) {
```
Would it make sense to define macros of `252` (max data length) and `3` below (header length?)

ntd replied:
It definitely does not hurt expressiveness: implemented in 651be816726ac9f77cc3d2837a77ec56e5022bfd.

jimklimov reviewed (commented):
For the first cursory look at this, looks quite solid - thanks! Feel welcome to update `NEWS.adoc` and perhaps `acknowledgements.txt`

Also a man page would be needed.

ntd commented:
> For the first cursory look at this, looks quite solid - thanks! Feel welcome to update `NEWS.adoc` and perhaps `acknowledgements.txt`
> 
> Also a man page would be needed.

First of all, many thanks for the review.

I just added manpage and updated NEWS.adoc. Not knowing where to put it (this is my first contribution to NUT), I added the announcement at the end of the `Release notes for NUT 2.8.3`. Feel free to move wherever more appropriate or drop me a line if I must do it.

ntd commented:
Rebased on c8d7e3c6164896d02de4dad8c03d6b772e7beb1c.

ntd commented:
After the manufacturer provided more info, I've been able to add a lot more stuff. Now I'm quite happy with the result.
I'll try to get Bicker involved in some way, at least to get some feedback.

jimklimov commented:
```
  CC       bicker_ser.o
[01m[K../../drivers/bicker_ser.c:[m[K In function '[01m[Kupsdrv_initinfo[m[K':
[01m[K../../drivers/bicker_ser.c:550:46:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  550 |                 dstate_setinfo("device.mfr", [01;31m[Kstring[m[K);
      |                                              [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:554:49:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  554 |                 dstate_setinfo("device.serial", [01;31m[Kstring[m[K);
      |                                                 [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:558:48:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  558 |                 dstate_setinfo("device.model", [01;31m[Kstring[m[K);
      |                                                [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:[m[K In function '[01m[Kupsdrv_initups[m[K':
[01m[K../../drivers/bicker_ser.c:717:48:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  717 |                 dstate_setinfo("ups.firmware", [01;31m[Kstring[m[K);
      |                                                [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:721:48:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  721 |                 dstate_setinfo("battery.type", [01;31m[Kstring[m[K);
      |                                                [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:728:52:[m[K [01;31m[Kerror: [m[Kformat not a string literal and no format arguments [[01;31m[K-Werror=format-security[m[K]
  728 |                 dstate_setinfo("ups.firmware.aux", [01;31m[Kstring[m[K);
      |                                                    [01;31m[K^~~~~~[m[K
[01m[K../../drivers/bicker_ser.c:[m[K At top level:
```

IIRC those should be `dstate_setinfo("some.name", "%s", stringvar);` wherever you have fixed strings provided by a third party (so they are not interpreted as a formatting pattern inadvertently). We do have a few cases of formatting strings being variables (coming from our tables) with macros hushing this warning, but that is a special case which needs extra care and supervision :)

ntd commented:
> IIRC those should be `dstate_setinfo("some.name", "%s", stringvar);` wherever you have fixed strings provided by a third party (so they are not interpreted as a formatting pattern inadvertently).

Of course you are right, sorry about that. I'm very well aware user provided strings in `printf` format arguments can lead to nasty bugs. I just fixed in c9fed627197ab43f62da675d26ae3c75b8d63f77.

jimklimov commented:
Thanks! And that reminded me to post a TODO #2450 :)

jimklimov commented:
Ugh, in `clang` scenarios above another concern has surfaced:
```
...
bicker_ser.c:129:23: note: expanded from macro 'SWAPONBE'
#define SWAPONBE(in)    (in)
                         ^~
bicker_ser.c:460:4: error: cast from 'uint8_t *' (aka 'unsigned char *') to 'uint16_t *' (aka 'unsigned short *') increases required alignment from 1 to 2 [-Werror,-Wcast-align]
        * (uint16_t *) &data[1] = SWAPONBE(parameter->val);
          ^~~~~~~~~~~~~~~~~~~~~
5 errors generated.
```

ntd commented:
Yes, I've seen that. I was fearing that could be an issue: I think the problem is not `SWAPONBE` itself but the calling code that is trying to access uint16_t on odd addresses. I'll try to come up with a solution. At the very least I can extract the uint16_t using mathematic.

jimklimov reviewed (changes requested):
Looks very solid, thank you again. I've got some suggestions for further perfection or double-checking, but I think none of those are critical/blockers.

jimklimov commented:
> Yes, I've seen that. I was fearing that could be an issue: I think the problem is not `SWAPONBE` itself but the calling code that is trying to access uint16_t on odd addresses. I'll try to come up with a solution. At the very least I can extract the uint16_t using mathematic.

I think you can try compiler hints to `align(1)` (don't remember the syntax OTOH) so it would not try to optimize memory access to the opaque buffer you are actually dealing with. Although I think it is implicit for a `char[]` buffer anyway. With a four-byte preamble (from buffer start to data) as per current offset definition, operations with the only two "data" bytes which are the 16-bit word would actually be aligned to whatever 1/2(/4?) byte alignment involved, so little hit performance-wise on platforms where getting to this starting data byte invokes extra CPU cycles. Dunno about 8-byte (64-bit) alignment though, might have some cost there.

By "extracting" you mean constructing a new `uint16_t` using byte selection and bit-shift - just without swapping them?

ntd commented:
> By "extracting" you mean constructing a new `uint16_t` using byte selection and bit-shift - just without swapping them?

Yes, I ended up doing that. The resulting code seems even cleaner and it does not rely on endianness anymore. I will push shortly after solving some of the issue you posted (and some cosmetic issues that are striking my eyes too).

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+/* For some reason the `seconds` delay (at least on my UPSIC-2403D)
+ * is not honored: the shutdown is always delayed by 2 seconds. This
+ * fixed delay seems to be independent from the state of the UPS (on
+ * line or on battery) and from the dip switches setting.
```
Fixed in beb26a68472e98992dec22db26044fbd747311d7.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+			TOUINT(buf[datalen + 4]), TOUINT(BICKER_EOT));
+		return -1;
+	} else if (idx != '\xEE' && buf[2] == '\xEE') {
+		/* I found sperimentally that, when the syntax is
```
Fixed in beb26a68472e98992dec22db26044fbd747311d7.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+	/* By default data is stored in little-endian so, on big-endian
+	 * platforms, a byte swap must be performed */
+	*dst = BYTESWAP(*dst);
+#endif
```
After c9ba796d336b9af44bb2c79665283328a6997bae this should be no more an issue.

ntd commented on NEWS.adoc:
```diff
@@ -146,6 +146,10 @@ https://github.com/networkupstools/nut/milestone/11
    respective warnings issued by the new generations of analysis tools.
    [#823, #2437, link:https://github.com/networkupstools/nut-website/issues/52[nut-website issue #52]]
 
+ - bicker_ser: added new driver for Bicker 12/24Vdc UPS via RS-232 serial
```
I copy&pasted the `adelsystem_cbi` announcement, that seems the more close to the `bicker_ser` description.

But now the protocol is based on the UPS-Gen² software user manual so, according to that manual, more UPSes are supported (even USB one):

![image](https://github.com/networkupstools/nut/assets/160618/9d469b59-5480-4087-b91d-57cd52a1a49b)

ntd replied:
So I think that description is at least incomplete. We'll see if Bicker has anything to say about that.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+	BickerParameter parameter;
+
+	upsfd = ser_open(device_path);
+	ser_set_speed(upsfd, device_path, B38400);
```
Yes, it is. The manual states that the parameter `'\x0A'` is a placeholder for RS232 port configuration. Not sure what that does mean apart you cannot change the baudrate.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+	ssize_t ret;
+	uint8_t response;
+
+	ret = bicker_send('\x03', '\x32', &seconds, 1);
```
Fixed in 0c80ece5927dcbab5e5609ba34aed38ca6853a1a.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+
+	/* Read the rest: command index (1 byte), command (1 byte), data
+	 * (datalen bytes) and EOT (1 byte), i.e. `datalen + 3` bytes */
+	ret = ser_get_buf_len(upsfd, buf + 2, datalen + 3, BICKER_TIMEOUT, 0);
```
Here 3 would ideally be `1 + BICKER_HEADER + 1  -  2` because it is the packet length - datalen - 2 (bytes already read). I admit that is confusing as hell.

Basically I read the packet in two steps: (1) the initial 2 bytes (where the size field resides) and (2) the rest of the packet. I tried to clarify the process in commit c78524c3f99b50ce3c08eedafb5bb0ac2eb0411e: let me know if that helps.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+				(int)datalen, BICKER_MAXDATA);
+			return -1;
+		}
+		memcpy(buf + 4, data, datalen);
```
Fixed in c78524c3f99b50ce3c08eedafb5bb0ac2eb0411e.

ntd commented on drivers/bicker_ser.c:
```diff
@@ -0,0 +1,751 @@
+	}
+
+	if (data != NULL) {
+		memcpy(data, buf + 4, datalen);
```
Fixed in c78524c3f99b50ce3c08eedafb5bb0ac2eb0411e.

jimklimov commented:
My condolences about the device... its sacrifice will be remembered :(
And I hope you find a way to un-brick it :)

jimklimov reviewed (approved):
Looks great and a lot better than much of the older code, I am out of ideas what else to complain about. Thank you very much for bearing with me and my nits :D

ntd commented on drivers/bicker_ser.c:
```diff
@@ -220,47 +232,48 @@ static ssize_t bicker_receive(char idx, char cmd, void *data)
-	ret = ser_get_buf_len(upsfd, buf + 2, datalen + 3, BICKER_TIMEOUT, 0);
+	/* Read the rest of the packet */
+	buflen = BICKER_PACKET(datalen);
+	ret = ser_get_buf_len(upsfd, buf + 2, buflen - 2, BICKER_TIMEOUT, 0);
```
Yes, exactly that in both cases.
